yarpl/include/yarpl/flowable/Flowable_FromObservable.h

```diff
@@ -40,58 +49,200 @@ class FlowableFromObservableSubscription
- private:
-  Reference<yarpl::observable::Observable<T>> observable_;
-  Reference<yarpl::flowable::Subscriber<T>> subscriber_;
-  Reference<yarpl::observable::Subscription> observableSubscription_;
+ protected:
+  virtual void onCreditsAvailable(int64_t /*credits*/) {}
+  virtual void onNextWithoutCredits(T /*t*/) {
+    // by default drop anything else received while we don't have credits
+  }
+
+  Reference<observable::Observable<T>> observable_;
+  Reference<flowable::Subscriber<T>> subscriber_;
   std::atomic_bool started{false};
   std::atomic<int64_t> requested_{0};
 };
+
+template <typename T>
+using FlowableFromObservableSubscriptionDropStrategy = FlowableFromObservableSubscription<T>;
+
+template <typename T>
+class FlowableFromObservableSubscriptionErrorStrategy : public FlowableFromObservableSubscription<T> {
+ using Super = FlowableFromObservableSubscription<T>;
+ public:
+  using Super::FlowableFromObservableSubscription;
+
+ private:
+  void onNextWithoutCredits(T /*t*/) override {
+    if(observable::Observer<T>::isUnsubscribedOrTerminated()) {
+      return;
+    }
+    Super::onError(MissingBackpressureException());
+    Super::cancel();
+  }
+};
+
+template <typename T>
+class FlowableFromObservableSubscriptionBufferStrategy : public FlowableFromObservableSubscription<T> {
+  using Super = FlowableFromObservableSubscription<T>;
+ public:
+  using Super::FlowableFromObservableSubscription;
+
+ private:
+  void onComplete() override {
+    if(!buffer_->empty()) {
+      // we have buffered some items so we will defer delivering on complete for
+      // later
+      completed_ = true;
+    } else {
+      Super::onComplete();
+    }
+  }
+
+  //
+  // onError signal is delivered immediately by design
+  //
+
+  void onNextWithoutCredits(T t) override {
+    if(Super::isUnsubscribed()) {
+      return;
+    }
+
+    buffer_->push_back(std::move(t));
+  }
+
+  void onCreditsAvailable(int64_t credits) override {
+    DCHECK(credits > 0);
+    auto&& lockedBuffer = buffer_.wlock();
+    while(credits-- > 0 && !lockedBuffer->empty()) {
+      Super::onNext(std::move(lockedBuffer->front()));
+      lockedBuffer->pop_front();
+    }
+
+    if(lockedBuffer->empty() && completed_) {
+      Super::onComplete();
+    }
+  }
+
+  folly::Synchronized<std::deque<T>> buffer_;
+  std::atomic<bool> completed_{false};
+};
+
+template <typename T>
+class FlowableFromObservableSubscriptionLatestStrategy : public FlowableFromObservableSubscription<T> {
+  using Super = FlowableFromObservableSubscription<T>;
+ public:
+  using Super::FlowableFromObservableSubscription;
+
+ private:
+  void onComplete() override {
+    if(storesLatest_) {
+      // we have buffered an item so we will defer delivering on complete for
+      // later
+      completed_ = true;
+    } else {
+      Super::onComplete();
+    }
+  }
+
+  //
+  // onError signal is delivered immediately by design
+  //
+
+  void onNextWithoutCredits(T t) override {
+    storesLatest_ = true;
```

Comment:
`latest_` being protected but `storesLatest_`  not seems racy

Reply:
storesLatest_ is atmoic... but I do agree that the model in general is broken..

Reply:
(y) that would be really great